### ISSUE TITLE
Implement TDA Metric

### ISSUE THREAD

**Issue by eddableheath** [alan-turing-institute/ARC-LoCoMoSeT#35]
Topological data analysis has a fairly well developed pipeline for these kinds of metrics. The paper mentioned in the state of work is on the more theoretical end, but according to Tom Davies there are multiple repos containing implementations of this pipeline.

Todo:

- [x] Talk with Tom to get recommendations of implementations and advice on usage.
- [x] Investigate which implementation is best/easiest to work with.
<del> - [ ] Implement within our framework. </del>

**Comment by eddableheath** [comment]
From chat with Tom:

What has become the defacto standard library for TDA is [Giotto-tda](https://giotto-ai.github.io/gtda-docs/0.5.1/index.html).

The pipe line would be as follows:

1. Use the [cubical homology](https://giotto-ai.github.io/gtda-docs/0.5.1/modules/generated/homology/gtda.homology.CubicalPersistence.html) to create a set of persistence diagrams from both the ideal case (just the data and labels), and the features post model inference with the labels.
2. Given each set of persistence diagrams create a [persistance image](https://giotto-ai.github.io/gtda-docs/0.5.1/modules/generated/diagrams/representations/gtda.diagrams.PersistenceImage.html) - essentially a vector representation of topological information.
3. Compare the (Euclidean) distance between the vector for each model and the vector for the base case, with the shorter the distance suggesting better performance post fine tuning of a model.

Tom makes the point that this use case for TDA is not unusual but would still be considered experimental. In light of that and the fact that this seems to be the most computationally expensive metric my approach is going to be one of trying to get the code working even if there are gaps in my understanding, before trying to get my head around it fully, in an effort to prototype it and sense test it.

**Comment by jack89roberts** [comment]
Sounds good, maybe limit it to say up to a couple of weeks' exploration initially, then at that point we can decide whether it's worth continuing with it or not.

**Comment by eddableheath** [comment]
Further detail from Tom:

- For homology -> diagram step we should give it either n (x_1, ..., x_n, y) or (x_1, ..., x_n, f(x)) tensors where x_i are the flattened pixel values of a grayscale version of image, y are the ground true labels, and f(x) are the features.
- The images need to be in grayscale for the homology analysis to work.
- The features can be from colour images (this should still work, but for somewhat unexact reasons).

**Comment by eddableheath** [comment]
Ok further details:

- The cubical homology isn't going to work as giotto-tda requires a numpy array of all the flattened pixel values. This will only work if we transform the incoming images first so they all have the same number of pixels, which is definitely an option but one I'm keeping it on  the backburner for now.
- The Vietoris Rips homology is simpler (can acccept ragged arrays) but a lot slower, trying this first.

Some general notes on how this method differs from all the others:

- The pre processing is different as we require a separate preprocessing of the images for the TDA comparison.
- This approach requires comparing all the models to be tested simultaneously to work (as the persistance image bounding box needs to consider all the diagrams it's comparing).

**Comment by eddableheath** [comment]
On a deeper inspection the Vietoris Rips homology also doesn't accept ragged arrays (somewhat misleading documentation...) so I will be looking at rescaling the images.

**Comment by eddableheath** [comment]
Current list of hyperparameters to experiment with:

1. Pixel number for image reduction: I've been using 224x224 as this is the standard that the preprocessors on Huggingface use, and this shouldn't change how effective the TDA is, however resizing them to n_classes x n_classes if n_classes > 224 could aid with the next step:
2. To append the label to the image a reasonable suggestion is to append a 'new row of pixel values' to the bottom. This could either be [ label ] * 224, or could be a one-hot vector. If a OHV then it would either need padding with zeros if n_classes < 224 or wrapping around and padding to add multiple rows. The alternative being that one can resize the images as above. Both methods would be reasonable, however the one hot method might make more topological sense.
3. When running inference through the model do we want the classes or the features? Again there doesn't seem to be a good immediate answer to this, however if we want features we'll likely need to reduce the dimensions and/or pad to make it all fit, and classes has the same question as (2), though which ever the answer to (2) is if we use classes it should correspond here.

I'm not sure (and also Tom isn't sure) the effect any of these difference should have on this pipeline overall other than they all seem like worthwhile solutions to problems so might be worth experimenting with all of them.

**Comment by eddableheath** [comment]
I've settled on the following hyperparameters to experiment with initially:

- number of images in the probe set.
- one hot vector vs. integer labels
- classifier (with one hot vectors or integer labels) vs. features

My initial experiments with the pipeline show that just using the classifier is not effective for ImageNet. This is to be expected as the models were trained on ImageNet and so are fairly accurate, in other words TDA is too blunt a tool to pick it apart in this case. However if we extract the features rather than the label we might see a difference.

I'll be running an experiment testing all of the above parameters for the most effective version, however will not implement the TDA metric within our structure until I've tested it as it will require a refactoring of our processes.

**Comment by eddableheath** [comment]
For the features I had been reducing the feature dimension down to number of pixels on one side of the squared images (224 as standard), and padding with zeros to get up to the correct dimension if the feature vector is smaller than the pixel number. However it's occured to me that a more topologically reasonably way to do this (i.e. avoiding an uneven proliferation of null dimensions for certain models - which would affect the outcome but not in a way that is representative of their performance necessarily) would be to feature reduce to a dimension that is some divisor of the pixel lenght (e.g. 28 or 56 for 224) and repeat the feature vector as an additional row of 'pixel values'.

**Comment by eddableheath** [comment]
Updated todo list:

- [x] Pull out functions into a script to run an experiment.
- [x] Write tests for functions.
- [x] Run experiment.
- [ ] Decide upon how to integrate into our previous framework, if at all.
- [ ] (following on from previous) refactor for non streamed data

**Comment by eddableheath** [comment]
I've been running an experiment where I vary the hyperparameters, and will have some results analysed by tomorrow. However I'm already convinced this isn't going to work as not only do the results I've been looking at not stack up but I'm also seeing it more to no longer being 'low-cost', with a single run taking up to three hours. Obviously there are some serious optimisations to make from my quite adhoc run however I don't see that there is enough to make it worth it.

**Comment by eddableheath** [comment]
From experiments this doesn't seem to work for ImageNet-1k, and moving forward it is unclear how one would use the labels in a sensible way for a dataset with different classes to ImageNet. TDA in this application appears to not be mature enough. Before closing this issue though the final three steps should be:

- [x] Check wih Tom that there aren't any obvious methods within TDA that I've missed for moving forward.
- [x] Discuss based on this information wheter or not it's worth moving forward with TDA.

**Comment by eddableheath** [comment]
Todo:

- [x] Write up in report why we're deciding to not move forward with TDA.

**Comment by eddableheath** [comment]
Closing this issue as TDA considered a dead end, as explained in report draf.
